Importer with bigtable support

replay opened the pull request:
Modifies the whisper-importer-reader and writer utilities so they:
1) use chunk write requests to send the data from the reader to the writer, which makes it easier to plug any store into the writer that satisfies the `mdata.Store` interface
2) support bigtable as a store and as an index, both can be chosen separately

Unfortunately some of the arguments that need to be passed to the writer had to be changed, so the invocation will need to be updated (HM-API).

It requires some of the data structures in github.com/raintank/schema to be msgp marshalable, which they currently aren't. Once this PR is accepted I'll create an according PR there.

This is not tested yet.

This already includes the PR https://github.com/grafana/metrictank/pull/1290, once that one is merged this diff will be shorter

woodsaj commented on cmd/mt-whisper-importer-reader/conversion.go:
```diff
@@ -51,17 +52,17 @@ func (c *conversion) getPoints(retIdx int, spp, nop uint32) map[string][]whisper
-		adjustedPoints["sum"] = make(map[uint32]float64)
+	adjustedPoints := make(map[schema.Method]map[uint32]float64)
+	if retIdx > 0 && c.method == schema.Avg || c.method == schema.Sum {
+		adjustedPoints[schema.Cnt] = make(map[uint32]float64)
```
why do we keep a count aggregation if all the user wants is 'sum'?

replay replied:
I think that was because that we can also serve average queries

woodsaj replied:
we should stop doing that.  The "count" value we are storing will be wrong any time the original metrics stream had null points.  eg, if the raw interval is 1min and their is a 10min SUM rollup, we cant assume that the sum is the result of 10 points.  It could be the result of anywhere from 0-10 points.

But maybe we should address this problem in a followup PR.

woodsaj commented on mdata/aggmetric.go:
```diff
@@ -356,7 +356,16 @@ func (a *AggMetric) persist(pos int) {
 	pending := make([]*ChunkWriteRequest, 1)
 	// add the current chunk to the list of chunks to send to the writeQueue
-	cwr := NewChunkWriteRequest(a, a.key, chunk, a.ttl, a.chunkSpan, time.Now())
+	cwr := NewChunkWriteRequest(func() {
```
Might be better to move the callback to just be `a.SyncChunkSaveState(chunk.Series.T0, true)`
and update SyncChunkSaveState to be a generator. eg.
```

// Generate the callback function for syncing chunk save state.
// 'ts' is the chunk T0 of the chunk that was saved, if sendPersist is true a 
// SendPersistMessage will be generated.
func (a *AggMetric) SyncChunkSaveState(ts uint32, sendPersist bool) ChunkSaveCallback {
	return func() {
		a.Lock()

		if ts > a.lastSaveFinish {
			a.lastSaveFinish = ts
		}
		if ts > a.lastSaveStart {
			a.lastSaveStart = ts
		}
		a.Unlock()
		log.Debugf("AM: metric %s at chunk T0=%d has been saved.", a.key, ts)
		if sendPersist {
			SendPersistMessage(a.key.String(), ts)
		}
	}
}
```

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,78 @@
+//go:generate msgp
 type ChunkWriteRequest struct {
-	Metric    *AggMetric
+	Callback  func() `msg:"-"`
```
We should define this as a type, rather then just using 'func()'. eg
```
type ChunkSaveCallback func()
```

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,78 @@
+// MdWithCwrs is used by the whisper importer utilities to transfer the data
+// from the reader to the writer
+//go:generate msgp
+type MdWithCwrs struct {
```
i do not like this struct name at all.    We want a struct to allow us to transport a batch of ChunkWriteRequests and the metadata needed to add an entry index.

I think the the old 'Archive' name was better. But perhaps we can use `ArchiveRequest`

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,78 @@
+// from the reader to the writer
+//go:generate msgp
+type MdWithCwrs struct {
+	Md   schema.MetricData
```
`Md` is obscure.  Please just call it `MetricData`

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,78 @@
+//go:generate msgp
+type MdWithCwrs struct {
+	Md   schema.MetricData
+	Cwrs []ChunkWriteRequest
```
`ChunkWriteRequests`

woodsaj commented on store/bigtable/bigtable.go:
```diff
@@ -269,9 +268,8 @@ func (s *Store) processWriteQueue(queue chan *mdata.ChunkWriteRequest, meter *st
-						buf[i].Metric.SyncChunkSaveState(buf[i].Chunk.Series.T0)
-						mdata.SendPersistMessage(buf[i].Key.String(), buf[i].Chunk.Series.T0)
-						log.Debugf("btStore: save complete. %s:%d %v", buf[i].Key, buf[i].Chunk.Series.T0, buf[i].Chunk)
+						buf[i].Callback()
```
`Callback()` should be optional, so check if it is nil before trying to call it.

woodsaj commented on store/cassandra/cassandra.go:
```diff
@@ -312,28 +312,26 @@ func (c *CassandraStore) processWriteQueue(queue chan *mdata.ChunkWriteRequest,
-					cwr.Metric.SyncChunkSaveState(cwr.Chunk.Series.T0)
-					mdata.SendPersistMessage(keyStr, cwr.Chunk.Series.T0)
-					log.Debugf("CS: save complete. %s:%d %v", keyStr, cwr.Chunk.Series.T0, cwr.Chunk)
+					cwr.Callback()
```
Callback() should be optional, so check if it is nil before trying to call it.

woodsaj commented on store/bigtable/bigtable.go:
```diff
@@ -291,9 +289,8 @@ func (s *Store) processWriteQueue(queue chan *mdata.ChunkWriteRequest, meter *st
-					cwr.Metric.SyncChunkSaveState(cwr.Chunk.Series.T0)
-					mdata.SendPersistMessage(cwr.Key.String(), cwr.Chunk.Series.T0)
-					log.Debugf("btStore: save complete. %s:%d %v", cwr.Key.String(), cwr.Chunk.Series.T0, cwr.Chunk)
+					cwr.Callback()
```
Callback() should be optional, so check if it is nil before trying to call it.

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,78 @@
+
+// MdWithCwrs is used by the whisper importer utilities to transfer the data
+// from the reader to the writer
+//go:generate msgp
```
We dont have to use msgp for this.   The majority of the data is already in []byte slices.  so there is not a large overhead in marshalling/unmarshalling.  We could just use `gob` if we wanted.

replay replied:
using msgp seems simple though. what would be the advantage of using gob? that we don't need to generate the msgp methods for the structs in github.com/raintank/schema ?

woodsaj replied:
basically.  Given msgp is already working, there is probably no point in replacing it.

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,78 @@
+	Cwrs []ChunkWriteRequest
+}
+
+func (m *MdWithCwrs) MarshalCompressed() (*bytes.Buffer, error) {
```
We need to version the payload, so that if we change how we encode data in future (eg removing gzip, or using a different compression algorithm) we can correctly handle it.

woodsaj commented on cmd/mt-whisper-importer-reader/main.go:
```diff
@@ -252,48 +252,47 @@ func sortPoints(points pointSorter) pointSorter {
-	res := archive.Metric{
-		AggregationMethod: uint32(w.Header.Metadata.AggregationMethod),
+func getMetric(w *whisper.Whisper, file, name string) (mdata.ArchiveRequest, error) {
+	res := mdata.ArchiveRequest{
```
we should move this to after checking that the whisper file has data.  No point in allocating memory if we dont need to.

woodsaj commented on cmd/mt-whisper-importer-reader/main.go:
```diff
@@ -304,52 +303,42 @@ func getMetric(w *whisper.Whisper, file, name string) (archive.Metric, error) {
 			}
-			mkey, err := schema.MKeyFromString(md.Id)
+
+			mkey, err := schema.MKeyFromString(res.MetricData.Id)
```
res.MetricData.Id doesnt change.  We should get the mkey once before looping through the retentions and points.

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,95 @@
+		return errors.New(fmt.Sprintf("ERROR: Creating Gzip reader: %q", err))
+	}
+
+	raw, err := ioutil.ReadAll(gzipReader)
```
There is no need to load the uncompressed data into memory, msgp can decode from a io.Reader directly.
eg,

```
err := msgp.Decode(gzipReader, a)
```

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,95 @@
+	// prefix the buffer with version number 1
+	buf.WriteByte(byte(uint8(1)))
+
+	b, err := a.MarshalMsg(nil)
```
No need to marshal to a bytes then pass the byte slice to gzip.  It can be done all at once.

```
err := msgp.Encode(gzip.NewWriter(&buf), a)
``

replay commented:
Thx for all the comments @woodsaj , will fix everything. What do you think about that I replaced the CLI flags with the same configuration method like what MT is using in https://github.com/grafana/metrictank/pull/1291/commits/84217dea0a7fb2eb7975b3925d8e574c60e1da25? I figured that way it's going to be much easier to configure it from HM-API, because the majority of the store/index configs are just going to be the same as in the MT write deployments.

woodsaj commented:
> What do you think about that I replaced the CLI flags with the same configuration method like what MT is using

I love it.

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,82 @@
+	versionBuf := make([]byte, 1)
+	readBytes, err := b.Read(versionBuf)
+	if err != nil || readBytes != 1 {
+		return errors.New(fmt.Sprintf("ERROR: Failed to read one byte: %s", err))
```
instead of `errors.New(fmt.Sprintf(...`  you can just use `fmt.Errorf(`

woodsaj commented on mdata/cwr.go:
```diff
@@ -1,23 +1,82 @@
+	if err != nil {
+		return errors.New(fmt.Sprintf("ERROR: Unmarshaling Raw: %q", err))
+	}
+
```
You need to close the gzipReader.
```
	if err = gzipReader.Close(); err != nil {
		return fmt.Errorf("ERROR: gzip reader error: %q", err)
	}
```

Dieterbe commented on mdata/aggmetric.go:
```diff
@@ -356,7 +361,14 @@ func (a *AggMetric) persist(pos int) {
+		a.key,
+		a.ttl,
+		chunk.Series.T0,
+		chunk.Encode(a.chunkSpan),
```
As i have mentioned before on slack, i'm a bit concerned about the extra encoding overhead in the critical - synchronous - section, which used to be in the async queue processing
unless there is any news that i missed, i think you should do a quick test using the docker stack and fakemetrics to confirm the difference in behavior with this patch applied under a significant chunk saving workload.

replay replied:
ok, i'll try that and compare if the ingest speed has changed due to that

replay replied:
I have filled 4h of fakemetrics into my kafka topic in the QA cluster, that kafka topic keeps the data for 7.7h, so then i had 3.7h of time to replay the contents of the topic multiple times with different images and compare how long the replaying takes. I did all those test runs with write pods, just manually killing them and then watching how long it takes for them to replay the kafka backlog.

I did 3 runs with each of `v0.11.0-295-gdd9b92d` and the latest image from this branch `whisper-importer-test8 (v0.11.0-297-gbb4a504)`
Unfortunately there seems to be quite a lot of variability between the different runs, so I think they might be unreliable and maybe I should better try comparing that on my local laptop or on bastion to get a more isolated environment.

```
us.gcr.io/metrictank-gcr/metrictank:v0.11.0-295-gdd9b92d:                                      

test1: 21:25:40 - 21:36:40 (11:00)
test2: 21:38:30 - 21:52:10 (13:40)
test3: 21:54:30 - 22:04:50 (10:20)
                                                                                                         
tus.gcr.io/metrictank-gcr/metrictank:whisper-importer-test8 (v0.11.0-297-gbb4a504):             

test1: 22:06:30 - 22:19:50 (13:20)
test2: 22:22:00 - 22:32:00 (12:00)
test3: 22:33:10 - 22:43:40 (10:30)
```

replay replied:
I've done another test on my local laptop, comparing the two branches. again, i filled kafka with a lot of data and made the different versions replay the data and insert it into cassandra. between each run i truncated all the cassandra tables. these are the results:

```
testing with latest master (v0.11.0-295-gdd9b92d)
run1: 13:56:54 - 14:02:42 (05:48)
run2: 14:11:33 - 14:17:18 (05:45)
run3: 15:06:49 - 15:12:55 (06:06)

testing with latest importer_bigtable (v0.11.0-298-gc2ba6e8)
run1: 14:33:00 - 14:38:43 (05:43)   
run2: 14:48:41 - 14:54:28 (05:47)   
run3: 14:56:04 - 15:02:11 (06:07)
```

Dieterbe replied:
can you share snapshots that are zoomed in on the timeframes where it creates all the cwr's ? i'm curious about the impact of the more expensive cwr creation.

replay replied:
I didn't capture snapshots, that means I'll need to do that again

Dieterbe replied:
i'm also curious to see if there's any change in memory stats

woodsaj replied:
I doubt there would be any negative CPU impact to encoding chunks before adding to the write queue.

with the encoding happening before we add to the write queue, we are constrained to 8 (number of threads processing ingested metrics) concurrent goroutines encoding chunks.   With chunks being encoded after being added to the write queue, there are 10 (default) write threads that can encode chunks at the same time.

There could be more memory used as the amount of time that both the *chunk.Chunk and []byte slice exist is greater.  but i wouldnt expect a significant impact from this.

replay replied:
I ran another benchmark. on my localhost i filled kafka with `3h` of data for `10000` metrics at `1sec` resolution. Then I started MT and made it write all that data to cassandra. Once I started the latest `master` and once this branch. Between each run I truncated all cassandra tables. 
The results are really very similar:

with importer_bigtable changes:
https://snapshot.raintank.io/dashboard/snapshot/IFx69Jh8yd1NGxy4ZBZCeCccNvW1iucS                                                        

with master
https://snapshot.raintank.io/dashboard/snapshot/jvNtJf0KoNkbYaNs3M1duCBTE5XKqFuM

replay commented:
@Dieterbe @woodsaj if you get a chance please take another look now, this should be ready to merge

Dieterbe commented on mdata/cwr.go:
```diff
@@ -1,23 +1,83 @@
+
+	version := uint8(versionBuf[0])
+	if version != 1 {
+		return errors.New(fmt.Sprintf("ERROR: Only version 1 is supported, received version %d", version))
```
fmt.Errorf

Dieterbe replied:
+ elsewhere

Dieterbe commented on mdata/cwr.go:
```diff
@@ -1,23 +1,83 @@
+	buf.WriteByte(byte(uint8(1)))
+
+	g := gzip.NewWriter(&buf)
+	msgp.Encode(g, a)
```
ignoring Encode error here?

Dieterbe commented on mdata/cwr.go:
```diff
@@ -1,23 +1,83 @@
+	return ChunkWriteRequest{callback, key, ttl, t0, data, ts}
+}
+
+// ArchiveRequest is used by the whisper importer utilities to transfer the data
```
types shouldn't declare which callers use them. this is bound to get out of date or incomplete quickly. if you want to know where a certain type is used and with much more accuracy, there's plenty of tools (such as go guru or vscode's find references)

Dieterbe replied:
if you want to document, better to document what the type is for. e.g. one could say it's a complete representation of a metric, both its timeseries data as well as its metadata.
the implications of how you could use it are then more obvious without tying us down.

Dieterbe commented on mdata/cwr_test.go:
```diff
@@ -0,0 +1,53 @@
+			{
+				TTL: 1,
+				T0:  2,
+			},
```
this test would be more useful if it actually verified that the `.Data` is correct as well

replay replied:
This test was broken and I didn't notice it. Apparently something weird happened when the cwr's `.Timestamp` was not set and got encoded/decoded. Once I've set it to any value the test passed.

replay commented on cmd/mt-whisper-importer-writer/main.go:
```diff
@@ -87,112 +59,100 @@ func init() {
-		flag.Usage()
+	config, err := globalconf.NewWithOptions(&globalconf.Options{
+		Filename:  path,
+		EnvPrefix: "IMPORTER_",
```
We'll probably want to be able to specify overrides which only apply to the importer, and not the MTs. That's the the main reason why I gave it a different prefix. 
That way HM-API can first generate all the `MT_` env vars as it would for an MT write deployment, search&replace `MT_` with `IMPORTER_` and then apply the `IMPORTER_` overrides.

Dieterbe replied:
wouldn't all the settings between importer and MT pods be largely the same? (same index and store configuration for example).
and if there are differences, well, they run in different pods, so they can't clobber one another.
using a different prefix seems to create more work for no reason that i can see (even you mention it would need custom search/replace code)

replay replied:
mostly the settings will be the same. but we might want to adjust stuff, like for example a different write concurrency or write queue size etc, so i just didn't want to make that impossible

Dieterbe replied:
again, different pods, so how would applying a different setting be impossible when using the same prefix?

replay replied:
> so how would applying a different setting be impossible when using the same prefix?

Because we only have one section for overrides in the whole instance config.

Alternatively, I could make the writer parse the variables with `MT_`, but in HM-API I allow the user to add overrides prefixed with `IMPORTER_` which would then be searched & replaced to `MT_` before the env vars get added to the importer deployment. Those two possible solutions seem about equally good to me.

Dieterbe replied:
aha so the clobbering is in the configuration itself, not in the application of it. makes sense.

All of the MT programs currently use the `MT_` prefix. 

```
grep -r EnvPrefix: cmd
cmd/mt-store-cat/main.go:		EnvPrefix: "MT_",
cmd/mt-kafka-mdm-sniff-out-of-order/main.go:		EnvPrefix: "MT_",
cmd/metrictank/metrictank.go:		EnvPrefix: "MT_",
cmd/mt-whisper-importer-writer/main.go:		EnvPrefix: "IMPORTER_",
cmd/mt-kafka-mdm-sniff/main.go:		EnvPrefix: "MT_",
```

I have a slight preference for consistency here (all using `MT_`) because it allows for easier config management and is simpler to work with.
The fact that hosted-metrics-api currently doesn't have a good way to specify different configuration for different tools or pods, is a problem that we can solve easily in hosted-metrics-api, and as you also mentioned, so let's do that.

replay replied:
ok, changing it to `MT_`

replay commented on cmd/mt-whisper-importer-writer/main.go:
```diff
@@ -87,112 +59,100 @@ func init() {
+	log.SetLevel(lvl)
+	log.Infof("logging level set to '%s'", *logLevel)
+
+	cluster.Init("mt-whisper-importer-writer", version, time.Now(), "http", int(80))
```
yes, it's exactly because of that, there was some `nil` pointer panic somewhere. I'll add a comment

replay commented on cmd/mt-whisper-importer-reader/conversion.go:
```diff
@@ -188,39 +191,41 @@ func decResolution(points []whisper.Point, method string, inRes, outRes, rawRes
-				Value:     (agg.Sum / agg.Cnt) * cnt,
-			})
-			agg.Reset()
-			return
```
This is clunky indeed, I didn't like it while i did this, at the same time I didn't want to mess with the constants of an external library.

replay replied:
way better now

replay commented:
I think I addressed all the feedback. I'll test it once more in QA, because since the last time I tested it there were quite a few changes.

replay commented:
I first redeployed my test instance in the QA cluster and made it use BigTable as index & store, then I imported 5 test metrics from generated whisper files, after restarting the read MT they showed up correctly.
Then I redeployed the instance and made it use Cassandra and imported the same test metrics into Cassandra, after restarting the read MT they again showed up correctly.

Dieterbe commented:
just fix my last comment,
and also why is anything changing in the vendor dir?
seems that in one place we removed an import for github.com/raintank/schema , andthen elsewhere introduced it again, so why are we making changes to it?

replay commented:
@Dieterbe regarding the changes in the vendor dir. because now we need the `schema.AMKey` and its member the `schema.Archive` to be marshallable (if that's a verb). I mentioned that in the top message: https://github.com/grafana/metrictank/pull/1291#issue-273243625
I'll create a PR to raintank/schema as soon as we're sure that this is the final solution that we want.

Dieterbe commented:
oh! well, when i came up with schema.Archive, MKey, Key, etc my idea was that these would all be internal implementation details that could be changed at any time because their exact values would only stay inside of a process, and never be serialized and sent in network requests.

For serialisation I think think we should use the string representation of the keys (e.g. 'orgid.deadbeef_sum_60'). this makes it easier to inspect them on the wire too

replay commented:
> For serialisation I think think we should use the string representation of the keys (e.g. 'orgid.deadbeef_sum_60').

That would mean we couldn't use `mdata.ChunkWriteRequest` anymore. We'd have to build some kind of struct which is basically the same, with the only difference being that it uses a string for the key. I thought one of the "nice" things in this PR is that it makes everything very short and concise, because we can directly transmit the cwrs, so we'd then lose that benefit.

Dieterbe commented:
> That would mean we couldn't use mdata.ChunkWriteRequest anymore. We'd have to build some kind of struct which is basically the same, with the only difference being that it uses a string for the key.

you're conflating a struct and its serialized representation.
you can totally use mdata.ChunkWriteRequest, but when serializing and deserializing, use the string representation of the AMKey. this may involve a tweak to AMKey to use its string representation in the msgp methods, i'm not sure what's the concrete way to do it.

replay commented:
@Dieterbe i think you're right, I'll try that

replay commented:
@Dieterbe how's this: https://github.com/grafana/metrictank/pull/1291/commits/39123b1b750303c84469d23e3ee16bf4acf6fed2
Or would you prefer if I try to find a way to implement the same thing without modifying raintank/schema at all? I think this would probably be possible by writing a similar extension for the ChunkWriteRequest type, it would just be a bit more complicated than this solution and I figured maybe somebody else also wants to encode the AMKey as string.

Dieterbe commented:
looks good, but there's 2 things left in schema that still need to be cleaned up:
Archive should not need any msgp code, and neither should Point, as far as i can see.

is it correct to say the only change in the schema library should be that we add msgp encoding to the AMKey (by serializing it to its string representation) ?

Also, the commit history has lots of back and forth, can we "compact" the commits (via git rebase -i)
it should ultimately be 2 or 3 commits:
1) the switch to using CWR for the importer, and all its corresponding changes in tests etc
2) the change in how we handle config management
3) (updates to docs/tools.md though that could be part of the 1/2 as well)

I suggest give it a shot, and see how far it can be cleaned up (i tend to do many runs of git rebase -i, each time squashing some commits together, until it becomes clear that i'll hit a merge conflict, or if i hit a merge conflict during rebasing i then just do `git rebase --abort`. if you do many smaller runs that still leaves you with the cleaned up history from previous runs)

replay commented:
`is it correct to say the only change in the schema library should be that we add msgp encoding to the AMKey (by serializing it to its string representation) ?`
The Archive and Key still need to have their generated msgp functions to encode/decode, because the generated msgp code for the cwrs expects them to be there, now they're just using that custom extension to do their job. F.e. this call: https://github.com/grafana/metrictank/blob/importer_bigtable/mdata/cwr_gen.go#L203

replay commented:
> is it correct to say the only change in the schema library should be that we add msgp encoding to the AMKey (by serializing it to its string representation) ?

I also fixed a typo in an error message and msgp generated a bunch of tests for that extension, otherwise that's right. If this PR is accepted I can create a PR to raintank/schema with those changes. There I can create separate commits for the typo and the msgp stuff. Then, once that's merged, we can just update the vendored `raintank/schema` to the latest master and there shouldn't be any code changes only the revision hash should change.

I fixed up the git history. Usually I do that by just squashing all changes into one commit, then I go back and edit that commit and use either `git add -p` or the `+` in vscode to add change by change and create separate commits that way (basically splitting it).
I made sure that even the first one of those two commits, without the second one, compiles. But it won't be useful, because it's not configurable. Only both together are useful.

Dieterbe commented:
so why does your PR still introduce a bunch of other files like point_gen_test.go and metric_gen_test.go ?

replay commented:
@Dieterbe sorry about that, that seems to be because i've generated the files by doing `go generate vendor/github.com/raintank/vendor/*.go` instead of just `key.go`. It's interesting though that there seems to be an "ungenerated" stringer directive, do we want that to be like this?

replay commented:
@Dieterbe I cleaned it up, please take another look

replay commented:
I found a critical bug and fixed it and added a test that covers it: 
https://github.com/grafana/metrictank/pull/1291/commits/0461e944a2d0640f16ef61e731506772a6194719#diff-2f8bddc453c2ba697972cbeff0844d5fL86

it took me forever to find that, i only saw that some metrics looked weird because certain chunks which were supposed to be archive chunks overwrote some raw chunks, but not always....

Dieterbe commented:
yep please patch in schema so we can merge there, introduce here cleanly, and merge here.
